[Android] Lexical model swapping + management

jahorton opened the pull request:
In the case that a user has two more more lexical models downloaded for the same language code, this PR allows proper management of their installation states.  Whenever the active model is changed, the previously-installed model will be de-installed automatically.  Afterward, the new model will be registered in the background and will be automatically activated once the Settings menus are closed.

This PR relies on #1836 at a minimum, though I've based it off of #1842 for ease of testing (and to facilitate a potential imminent demo).  The recent longpress PR's changes (#1844) also appear here since the longpress changes haven't yet been merged into the base.  (Basically - ignore KMKeyboard.java's changes.)

keyman-server commented:
[OSK renders for Windows](https://build.palaso.org/repository/download/KeymanDesktop_TestPrRenderOnScreenKeyboards/155027:id/kvk/index.html)

mcdurdin commented on android/KMEA/app/src/main/java/com/tavultesoft/kmea/KeyboardPickerActivity.java:
```diff
@@ -486,18 +486,21 @@ protected static boolean removeLexicalModel(Context context, int position) {
     if (result) {
-      Toast.makeText(context, "Model deleted", Toast.LENGTH_SHORT).show();
+      if(!silenceNotification) {
+        Toast.makeText(context, "" + "Model deleted", Toast.LENGTH_SHORT).show();
```
Looking good apart from using the term "Model" where we should be saying "Dictionary" in the UI :)

Do we have any resource strings in the Android UI? If so, is it possible for us to avoid adding extra UI text into the code which will make future localisation more tedious?

mcdurdin commented:
Crash when touching an installed language model in the Language settings view:

```
2019-06-26 19:46:39.089 1853-2083/com.tavultesoft.kmapro.debug E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
    Process: com.tavultesoft.kmapro.debug, PID: 1853
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:353)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ProgressDialog.setMax(int)' on a null object reference
        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:570)
        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:495)
        at android.os.AsyncTask$2.call(AsyncTask.java:333)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764) 
2019-06-26 19:46:39.346 1853-2083/com.tavultesoft.kmapro.debug E/UncaughtException: java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:353)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ProgressDialog.setMax(int)' on a null object reference
        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:570)
        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:495)
        at android.os.AsyncTask$2.call(AsyncTask.java:333)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764) 
```

mcdurdin commented:
I think this happened when Keyman was checking for keyboard/model updates (perhaps caused by being offline?):

```
2019-06-26 20:03:30.600 3229-4079/com.tavultesoft.kmapro.debug E/AndroidRuntime: FATAL EXCEPTION: AsyncTask #1
    Process: com.tavultesoft.kmapro.debug, PID: 3229
    java.lang.RuntimeException: An error occurred while executing doInBackground()
        at android.os.AsyncTask$3.done(AsyncTask.java:353)
        at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:383)
        at java.util.concurrent.FutureTask.setException(FutureTask.java:252)
        at java.util.concurrent.FutureTask.run(FutureTask.java:271)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
        at java.lang.Thread.run(Thread.java:764)
     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ProgressDialog.setMax(int)' on a null object reference
        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:630)
        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:542)
        at android.os.AsyncTask$2.call(AsyncTask.java:333)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:245) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636) 
        at java.lang.Thread.run(Thread.java:764) 
```

mcdurdin commented:
The following may be a secondary exception caused by the first (unclear to me at present):

```
2019-06-26 20:03:32.216 3229-3229/com.tavultesoft.kmapro.debug E/WindowManager: android.view.WindowLeaked: Activity com.tavultesoft.kmea.LanguagesSettingsActivity has leaked window DecorView@11f1758[] that was originally added here
        at android.view.ViewRootImpl.<init>(ViewRootImpl.java:494)
        at android.view.WindowManagerGlobal.addView(WindowManagerGlobal.java:346)
        at android.view.WindowManagerImpl.addView(WindowManagerImpl.java:93)
        at android.app.Dialog.show(Dialog.java:330)
        at com.tavultesoft.kmea.LanguagesSettingsActivity$2.onClick(LanguagesSettingsActivity.java:141)
        at android.view.View.performClick(View.java:6311)
        at android.view.View$PerformClick.run(View.java:24996)
        at android.os.Handler.handleCallback(Handler.java:790)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6537)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:857)
```

mcdurdin commented:
Various errors occur just after installing Keyman, in the web view; some of these may be a cascade:

```
VM16:1 Uncaught ReferenceError: updateKMText is not defined
    at <anonymous>:1:1
VM17:1 Uncaught ReferenceError: updateKMSelectionRange is not defined
    at <anonymous>:1:1
keyman.js:20030 Uncaught TypeError: Cannot read property 'adjustHeights' of undefined
    at KeymanBase.keymanweb.correctOSKTextSize (keyman.js:20030)
    at setOskHeight (keyboard.html:60)
    at window.onload (<anonymous>:1:48)
```

mcdurdin commented:
> I think this happened when Keyman was checking for keyboard/model updates (perhaps caused by being offline?):

Verified that as far as I can tell, this error only occurs when the phone is offline and is related to checking for updates.

jahorton commented:
Okay, so on the first:

>      Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ProgressDialog.setMax(int)' on a null object reference
>        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:570)
>        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:495)

Second one:

>     Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'void android.app.ProgressDialog.setMax(int)' on a null object reference
>        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:630)
>        at com.tavultesoft.kmea.data.CloudRepository$CloudDownloadTask.doInBackground(CloudRepository.java:542)

Different line numbers, but the method call behind the error is the same.  OK, gives me something to look at.

jahorton commented:
OK, I've addressed most the concerns brought up by those error reports and a few extra things.  Granted, I couldn't get an exact repro on the following:

> Various errors occur just after installing Keyman, in the web view; some of these may be a cascade:

But that did lead me to find a number of other first-time install issues that are now resolved by this PR, which should make for a smoother experience and a better first impression.

While I can see the `android.view.WindowLeaked` exception in the logs, I've currently got little idea what causes it, and it doesn't seem to affect the app's actual stability.  It mentions a `DecorView`, but that's only referenced by `KMTextView` rather than by the settings UI.  Not sure what's up there.
